python/tvm/relay/op/strategy/generic.py

```diff
@@ -1760,6 +1769,44 @@ def cumsum_strategy(attrs, inputs, out_type, target):
     return strategy
 
 
+@override_native_generic_func("concat_strategy")
+def concatenate_strategy(attrs, inputs, out_type, target):
+    """concatenate generic strategy"""
+    strategy = _op.OpStrategy()
+    strategy.add_implementation(
+        wrap_compute_concat(topi.concatenate),
+        wrap_topi_schedule(topi.generic.schedule_extern),
+        name="concatenate",
+    )
+    return strategy
+
+
+@concatenate_strategy.register(["cpu"])
+def concatenate_strategy_cpu(attrs, inputs, out_type, target):
+    """concatenate x86 strategy"""
+    strategy = _op.OpStrategy()
+    use_old_concat = False
+    for inpt in inputs:
+        shape = inpt.shape
+        for i in shape:
+            if not isinstance(i, tir.expr.IntImm):
+                use_old_concat = True
+                break
+    if use_old_concat:
+        strategy.add_implementation(
+            wrap_compute_concat(topi.transform.concatenate),
+            wrap_topi_schedule(topi.x86.injective.schedule_concatenate),
+            name="concatenate.generic",
+        )
+    else:
+        strategy.add_implementation(
+            wrap_compute_concat(topi.x86.concatenate),
+            wrap_topi_schedule(topi.x86.schedule_concatenate_cpu),
+            name="concatenate.cpu",
+        )
```

Comment:
It looks like you want to use the new impl by default unless the input has dynamic shape. Have you benchmarked that the new impl is always faster?

A better way, as we discussed before, is to let autotvm pick the better one based on tuning results. You need to add two implementations to the strategy, and choose the default one. That shouldn't be a controversial change.

Reply:
1. Regarding the dynamic shaping: the new version can be adapted if "any_val" is not presented within the shape.
2. Yes I tried to do benchmarking and it was faster in most cases. The problems may happen if available threads number is high because the "complexity" of solution is small in this case.
3. I Updated code to have two strategies